cudax/include/cuda/experimental/__container/async_buffer.cuh

```diff
@@ -0,0 +1,858 @@
+  {
+    const auto __count = static_cast<size_type>(__ilist.size());
+    this->__assign_impl(__count, __ilist.begin(), __ilist.end());
+  }
+
+  //! @brief Replaces the content of the async_buffer with the range \p __range
+  //! @param __range The range to be copied into this async_buffer.
+  //! @note Neither frees not allocates memory if `__range.size() == 0`.
+  _CCCL_TEMPLATE(class _Range)
+  _CCCL_REQUIRES(__compatible_range<_Range> _CCCL_AND _CUDA_VRANGES::forward_range<_Range> _CCCL_AND
+                   _CUDA_VRANGES::sized_range<_Range>)
+  _CCCL_HIDE_FROM_ABI void assign_range(_Range&& __range)
+  {
+    const auto __count = _CUDA_VRANGES::size(__range);
+    using _Iter        = _CUDA_VRANGES::iterator_t<_Range>;
+    this->__assign_impl<_Iter, __detect_transfer_kind<__is_host_only, _Range>>(
+      __count, _CUDA_VSTD::begin(__range), _CUDA_VRANGES::__unwrap_end(__range));
+  }
+
+#ifndef _CCCL_DOXYGEN_INVOKED // doxygen conflates the overloads
+  //! @brief Replaces the content of the async_buffer with the range \p __range
+  //! @param __range The range to be copied into this async_buffer.
+  //! @note Neither frees nor allocates memory if `__range.size() == 0`.
+  _CCCL_TEMPLATE(class _Range)
+  _CCCL_REQUIRES(__compatible_range<_Range> _CCCL_AND _CUDA_VRANGES::forward_range<_Range> _CCCL_AND(
+    !_CUDA_VRANGES::sized_range<_Range>))
+  _CCCL_HIDE_FROM_ABI void assign_range(_Range&& __range)
+  {
+    const auto __first = _CUDA_VRANGES::begin(__range);
+    const auto __last  = _CUDA_VRANGES::__unwrap_end(__range);
+    const auto __count = static_cast<size_type>(_CUDA_VRANGES::distance(__first, __last));
+
+    using _Iter = _CUDA_VRANGES::iterator_t<_Range>;
+    this->__assign_impl<_Iter, __detect_transfer_kind<__is_host_only, _Range>>(__count, __first, __last);
+  }
+#endif // _CCCL_DOXYGEN_INVOKED
+  //! @}
+
+  //! @brief Swaps the contents of a async_buffer with those of \p __other
+  //! @param __other The other async_buffer.
+  _CCCL_HIDE_FROM_ABI void swap(async_buffer& __other) noexcept
+  {
+    _CUDA_VSTD::swap(__buf_, __other.__buf_);
+    _CUDA_VSTD::swap(__size_, __other.__size_);
+  }
+
+  //! @brief Swaps the contents of two async_buffers
+  //! @param __lhs One async_buffer.
+  //! @param __rhs The other async_buffer.
+  _CCCL_HIDE_FROM_ABI friend void swap(async_buffer& __lhs, async_buffer& __rhs) noexcept
+  {
+    __lhs.swap(__rhs);
+  }
+  //! @}
+
+  //! @addtogroup comparison
+  //! @{
+
+  //! @brief Compares two async_buffers for equality
+  //! @param __lhs One async_buffer.
+  //! @param __rhs The other async_buffer.
+  //! @return true, if \p __lhs and \p __rhs contain equal elements have the same size
+  _CCCL_NODISCARD_FRIEND _CCCL_HIDE_FROM_ABI bool
+  operator==(const async_buffer& __lhs, const async_buffer& __rhs) noexcept(noexcept(_CUDA_VSTD::equal(
+    __lhs.__unwrapped_begin(), __lhs.__unwrapped_end(), __rhs.__unwrapped_begin(), __rhs.__unwrapped_end())))
```

Comment:
neither `cuda::std::equal` nor `thrust::equal` are ever `noexcept`

Reply:
Dropped the noexcept